src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -64,11 +78,328 @@
+
+
+def _get_first_argument_type(node) -> DataType:
+    """Helper function for the common IntrinsicCall case where
+    the return type matches exactly the datatype of the first argument.
+
+    :param node: The IntrinsicCall whose return type to compute.
+    :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+    :returns: the datatype of the first argument of the IntrinsicCall.
+    """
+    return node.arguments[0].datatype
+
+
+def _get_first_argument_type_with_optional_kind(node) -> DataType:
+    """Helper function for the common IntrinsicCall case where the
+    return type is the Intrinsic of the first argument, with an optional
+    kind parameter which may override the precision.
+
+    :param node: The IntrinsicCall whose return type to compute.
+    :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+    :returns: the datatype of the first argument of the IntrinsicCall.
+    """
+    if "kind" not in node.argument_names:
+        return node.arguments[0].datatype
+    else:
+        kind = node.arguments[node.argument_names.index("kind")]
+        return_type = node.arguments[0].datatype.copy()
+        return_type._precision = kind
+        return return_type
+
+
+def _get_first_argument_intrinsic_with_optional_kind_and_dim(node) -> DataType:
+    """Helper function for IntrinsicCalls like MAXLOC where they have optional
+    Kind and Dim options but the intrinsic is that of the first argument.
+
+    :param node: The IntrinsicCall whose return type to compute.
+    :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+    :returns: the computed datatype for the IntrinsicCall.
+    """
+    if "kind" in node.argument_names:
+        dtype = ScalarType(
+            node.arguments[0].datatype.datatype.intrinsic,
+            node.arguments[node.argument_names.index("kind")],
+        )
+    else:
+        # PSyclone has the UNDEFINED Precision as the default kind for all
+        # supported inbuilt datatypes.
+        dtype = ScalarType(
+            node.arguments[0].datatype.intrinsic,
+            ScalarType.Precision.UNDEFINED,
+        )
+    if "dim" not in node.argument_names:
+        return ArrayType(
+            dtype,
+            [
+                ArrayType.ArrayBounds(
+                    Literal("1", INTEGER_TYPE),
+                    Literal(
+                        str(len(node.arguments[0].datatype.shape)),
+                        INTEGER_TYPE,
+                    ),
+                )
+            ],
+        )
+    # Always have dim from here.
+    # If array has rank 1, the result is scalar.
+    arg = node.arguments[0]
+    shape = arg.datatype.shape
+    if len(shape) == 1:
+        return dtype
+    # For now we don't attempt to work out the shape.
+    new_shape = [ArrayType.Extent.DEFERRED] * (len(shape) - 1)
+    return ArrayType(dtype, new_shape)
+
+
+def _get_first_argument_logical_kind_with_optional_dim(node) -> DataType:
```

Comment:
Is it possible to replace the various `_get_first_argument_{int,real,logical...}_kind_with...` with a version that just takes the intrinsic-type as an argument?